File: backend-java/src/main/java/com/aisuite/service/LlmClient.java
package com.aisuite.service;

import com.aisuite.config.AiSuiteProperties;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
public class LlmClient {
    private static final Logger log = LoggerFactory.getLogger(LlmClient.class);
    private static final Set<String> PLACEHOLDER_KEYS = Set.of("", "sk-your-key-here", "your-api-key", "sk-xxx");

    private final AiSuiteProperties props;
    private final ObjectMapper mapper;
    private final HttpClient http;
    private final boolean apiAvailable;
    private String lastError;

    public LlmClient(AiSuiteProperties props, ObjectMapper mapper) {
        this.props = props;
        this.mapper = mapper;
        this.http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(30)).build();
        String key = props.getOpenaiApiKey() == null ? "" : props.getOpenaiApiKey().trim();
        this.apiAvailable = !key.isEmpty() && !PLACEHOLDER_KEYS.contains(key);
    }

    public boolean isAvailable() { return apiAvailable; }
    public String getLastError() { return lastError; }

    public String chat(List<Map<String, String>> messages, double temperature, boolean jsonMode, boolean useModel) {
        if (!useModel || !apiAvailable) return fallback(messages);
        try {
            Map<String, Object> body = new LinkedHashMap<>();
            body.put("model", props.getOpenaiModel());
            body.put("messages", messages);
            body.put("temperature", temperature);
            if (jsonMode) body.put("response_format", Map.of("type", "json_object"));

            String payload = mapper.writeValueAsString(body);
            HttpRequest req = HttpRequest.newBuilder()
                    .uri(URI.create(props.getOpenaiBaseUrl().replaceAll("/$", "") + "/chat/completions"))
                    .timeout(Duration.ofSeconds(90))
                    .header("Authorization", "Bearer " + props.getOpenaiApiKey())
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(payload))
                    .build();

            HttpResponse<String> res = http.send(req, HttpResponse.BodyHandlers.ofString());
            if (res.statusCode() >= 400) {
                lastError = res.body();
                log.warn("LLM API failed: {}", lastError);
                return fallback(messages);
            }
            JsonNode root = mapper.readTree(res.body());
            return root.path("choices").path(0).path("message").path("content").asText("");
        } catch (Exception e) {
            lastError = e.getMessage();
            log.warn("LLM API failed, using fallback: {}", e.getMessage());
            return fallback(messages);
        }
    }

    private String fallback(List<Map<String, String>> messages) {
        String userText = "";
        for (int i = messages.size() - 1; i >= 0; i--) {
            if ("user".equals(messages.get(i).get("role"))) {
                userText = extractUserText(messages.get(i).get("content"));
                break;
            }
        }
        String systemText = messages.stream()
                .filter(m -> "system".equals(m.get("role")))
                .map(m -> m.get("content"))
                .findFirst().orElse("");

        try {
            if (systemText.toLowerCase().contains("json") || systemText.contains("JSON")) {
                if (isSpeakingPrompt(systemText)) {
                    var corrections = demoGrammarCorrections(userText);
                    Map<String, Object> out = new LinkedHashMap<>();
                    if (!corrections.isEmpty()) {
                        out.put("reply", "Good effort! I noticed a few grammar points — check the corrections below. Could you try saying that again?");
                        out.put("encouragement", "Nice try! Applying corrections will make you sound more natural.");
                        out.put("pronunciation_score", 72);
                    } else {
                        out.put("reply", "That's a great point! Could you tell me more about your experience?");
                        out.put("encouragement", "Keep going — your fluency is improving!");
                        out.put("pronunciation_score", 78);
                    }
                    out.put("corrections", corrections);
                    out.put("pronunciation_feedback", "Clear articulation. Try stressing key verbs more.");
                    return mapper.writeValueAsString(out);
                }
                if (systemText.toLowerCase().contains("english coach")) {
                    return mapper.writeValueAsString(Map.of(
                            "overall_score", 75,
                            "strengths", List.of("词汇量不错", "语速自然"),
                            "improvements", List.of("多使用连接词", "注意动词时态"),
                            "key_phrases_learned", List.of("in my opinion", "as a result"),
                            "next_steps", "建议在同一场景下练习更难的追问。"
                    ));
                }
            }
        } catch (Exception e) {
            return "{\"reply\":\"That's interesting! Could you tell me more?\"}";
        }
        return "That's interesting! Could you tell me more about that?";
    }

    private boolean isSpeakingPrompt(String systemText) {
        String s = systemText.toLowerCase();
        return s.contains("interviewer") || s.contains("waiter") || s.contains("project manager")
                || s.contains("conversation") || s.contains("respond in json");
    }

    private String extractUserText(String content) {
        String[] lines = content.split("\n");
        for (int i = lines.length - 1; i >= 0; i--) {
            if (lines[i].toLowerCase().startsWith("user:")) {
                return lines[i].substring(lines[i].indexOf(':') + 1).trim();
            }
        }
        return content.trim();
    }

    private List<Map<String, String>> demoGrammarCorrections(String text) {
        record Rule(String pattern, String replacement, String reason, String severity) {}
        List<Rule> rules = List.of(
                new Rule("\\bI am work\\b", "I work", "Use simple present: I work / I am working", "major"),
                new Rule("\\bfor (\\w+) year\\b", "for $1 years", "Countable noun 'year' needs plural", "minor"),
                new Rule("\\bHe don't\\b", "He doesn't", "Third-person singular uses doesn't", "major"),
                new Rule("\\bit are\\b", "it is", "Subject-verb agreement: it is", "major"),
                new Rule("\\bI have went\\b", "I went", "Use past simple for finished past actions", "major"),
                new Rule("\\bdont\\b", "don't", "Missing apostrophe in contraction", "minor"),
                new Rule("\\bcoffe\\b", "coffee", "Spelling: coffee", "minor"),
                new Rule("\\brecieved\\b", "received", "Spelling: received", "minor"),
                new Rule("\\bmessege\\b", "message", "Spelling: message", "minor"),
                new Rule("\\breponse\\b", "response", "Spelling: response", "minor"),
                new Rule("\\bto learning\\b", "to learn", "After 'to' use base verb: learn", "major"),
                new Rule("\\bdiscuss about\\b", "discuss", "Discuss is transitive; omit 'about'", "minor"),
                new Rule("\\bmore better\\b", "better", "Better already means 'more good'", "minor")
        );
        List<Map<String, String>> corrections = new ArrayList<>();
        Set<String> seen = new HashSet<>();
        for (Rule rule : rules) {
            Pattern p = Pattern.compile(rule.pattern(), Pattern.CASE_INSENSITIVE);
            Matcher m = p.matcher(text);
            if (!m.find()) continue;
            String original = m.group();
            if (seen.contains(original.toLowerCase())) continue;
            seen.add(original.toLowerCase());
            Matcher single = p.matcher(original);
            String suggestion = single.replaceFirst(rule.replacement());
            corrections.add(Map.of(
                    "original", original,
                    "suggestion", suggestion,
                    "reason", rule.reason(),
                    "severity", rule.severity()
            ));
        }
        return corrections;
    }

    public JsonNode parseJson(String raw) {
        try {
            return mapper.readTree(raw);
        } catch (Exception e) {
            int start = raw.indexOf('{');
            int end = raw.lastIndexOf('}');
            if (start >= 0 && end > start) {
                try { return mapper.readTree(raw.substring(start, end + 1)); } catch (Exception ignored) {}
            }
            return mapper.createObjectNode().put("reply", raw);
        }
    }
}

File: backend/app/services/llm_client.py
"""Unified LLM client with OpenAI API and offline fallback."""

from __future__ import annotations

import json
import re
from typing import Any

import logging

from openai import APIConnectionError, APIStatusError, APITimeoutError, AsyncOpenAI

from app.config import settings

logger = logging.getLogger(__name__)

PLACEHOLDER_KEYS = {"", "sk-your-key-here", "your-api-key", "sk-xxx"}


class LLMClient:
    def __init__(self) -> None:
        self._client: AsyncOpenAI | None = None
        self._use_api = False
        self._last_error: str | None = None
        key = (settings.openai_api_key or "").strip()
        if key and key not in PLACEHOLDER_KEYS:
            self._client = AsyncOpenAI(
                api_key=key,
                base_url=settings.openai_base_url,
                timeout=90.0,
                max_retries=1,
            )
            self._use_api = True

    @property
    def available(self) -> bool:
        return self._use_api

    async def chat(
        self,
        messages: list[dict[str, str]],
        *,
        temperature: float = 0.7,
        json_mode: bool = False,
        use_model: bool = True,
    ) -> str:
        if not use_model or not self._client:
            return self._fallback(messages)

        kwargs: dict[str, Any] = {
            "model": settings.openai_model,
            "messages": messages,
            "temperature": temperature,
        }
        if json_mode:
            kwargs["response_format"] = {"type": "json_object"}

        try:
            response = await self._client.chat.completions.create(**kwargs)
            return response.choices[0].message.content or ""
        except (APITimeoutError, APIConnectionError, APIStatusError) as exc:
            logger.warning("LLM API failed, using fallback: %s", exc)
            self._last_error = str(exc)
            return self._fallback(messages)

    def _extract_user_text(self, content: str) -> str:
        return self._extract_last_user_turn(content)

    def _extract_last_user_turn(self, content: str) -> str:
        lines = content.strip().splitlines()
        for line in reversed(lines):
            if line.lower().startswith("user:"):
                return line.split(":", 1)[1].strip()
        return content.strip()

    def _demo_grammar_corrections(self, text: str) -> list[dict[str, str]]:
        rules: list[tuple[str, str, str, str]] = [
            (r"\bI am work\b", "I work", "Use simple present: I work / I am working", "major"),
            (r"\bfor (\w+) year\b", r"for \1 years", "Countable noun 'year' needs plural", "minor"),
            (r"\bHe don't\b", "He doesn't", "Third-person singular uses doesn't", "major"),
            (r"\bShe don't\b", "She doesn't", "Third-person singular uses doesn't", "major"),
            (r"\bIt don't\b", "It doesn't", "Third-person singular uses doesn't", "major"),
            (r"\bit are\b", "it is", "Subject-verb agreement: it is", "major"),
            (r"\bthey is\b", "they are", "Subject-verb agreement: they are", "major"),
            (r"\bI have went\b", "I went", "Use past simple for finished past actions", "major"),
            (r"\bhave went\b", "went", "Use past simple: went (not have went)", "major"),
            (r"\band buy\b", "and bought", "Keep past tense consistent: bought", "major"),
            (r"\bmany souvenir\b", "many souvenirs", "Countable nouns need plural form", "minor"),
            (r"\ba informations\b", "information", "Information is uncountable", "major"),
            (r"\bmore better\b", "better", "Better already means 'more good'", "minor"),
            (r"\bdiscuss about\b", "discuss", "Discuss is transitive; omit 'about'", "minor"),
            (r"\bdont\b", "don't", "Missing apostrophe in contraction", "minor"),
            (r"\bcoffe\b", "coffee", "Spelling: coffee (double f, double e)", "minor"),
            (r"\brecieved\b", "received", "Spelling: i before e except after c", "minor"),
            (r"\bmessege\b", "message", "Spelling: message", "minor"),
            (r"\breponse\b", "response", "Spelling: response", "minor"),
            (r"\bto learning\b", "to learn", "After 'to' use base verb: learn", "major"),
            (r"\bvery much like to\b", "really like to", "More natural word order", "minor"),
        ]

        corrections: list[dict[str, str]] = []
        seen: set[str] = set()
        for pattern, replacement, reason, severity in rules:
            match = re.search(pattern, text, flags=re.IGNORECASE)
            if not match:
                continue
            original = match.group(0)
            if original.lower() in seen:
                continue
            seen.add(original.lower())
            suggestion = re.sub(pattern, replacement, original, count=1, flags=re.IGNORECASE)
            corrections.append(
                {
                    "original": original,
                    "suggestion": suggestion,
                    "reason": reason,
                    "severity": severity,
                }
            )
        return corrections

    def _fallback(self, messages: list[dict[str, str]]) -> str:
        """Rule-based fallback when no API key is configured."""
        user_text = ""
        for msg in reversed(messages):
            if msg.get("role") == "user":
                user_text = self._extract_user_text(msg.get("content", ""))
                break

        system_text = next(
            (m.get("content", "") for m in messages if m.get("role") == "system"),
            "",
        )

        if "JSON" in system_text or "json" in system_text:
            if (
                "interviewer" in system_text.lower()
                or "waiter" in system_text.lower()
                or "project manager" in system_text.lower()
                or "conversation" in system_text.lower()
                or "respond in json" in system_text.lower()
            ):
                corrections = self._demo_grammar_corrections(user_text)
                if corrections:
                    reply = (
                        "Good effort! I noticed a few grammar points — check the corrections below. "
                        "Could you try saying that again?"
                    )
                    encouragement = "Nice try! Applying corrections will make you sound more natural."
                else:
                    reply = "That's a great point! Could you tell me more about your experience?"
                    encouragement = "Keep going — your fluency is improving!"
                return json.dumps(
                    {
                        "reply": reply,
                        "corrections": corrections,
                        "pronunciation_score": 78 if not corrections else 72,
                        "pronunciation_feedback": "Clear articulation. Try stressing key verbs more.",
                        "encouragement": encouragement,
                    },
                    ensure_ascii=False,
                )
            if "english coach" in system_text.lower() or "summary" in system_text.lower():
                return json.dumps(
                    {
                        "overall_score": 75,
                        "strengths": ["词汇量不错", "语速自然"],
                        "improvements": ["多使用连接词", "注意动词时态"],
                        "key_phrases_learned": ["in my opinion", "as a result"],
                        "next_steps": "建议在同一场景下练习更难的追问。",
                    },
                    ensure_ascii=False,
                )

        return "That's interesting! Could you tell me more about that?"


llm_client = LLMClient()
